Run soltest and cli tests on Linux ARM

matheusaaguiar opened the pull request:
Currently we have ARM releases but we only run bytecode equivalence tests.
This PR adds a job in the CI to run `soltest` and cli tests on ARM binary.
Closes #11351.
Closes #16321.

~~depends on #16328.~~

matheusaaguiar committed:
run soltest and cli tests on arm

cameel commented on .circleci/config.yml:
```diff
@@ -1987,8 +1998,10 @@ workflows:
+      - t_ubu_arm_soltest_all: *requires_b_ubu_static_arm
       - b_ubu_clang: *requires_nothing
       - t_ubu_clang_soltest: *requires_b_ubu_clang
+      - t_ubu_arm_soltest_all: *requires_b_ubu_static_arm
```
You have soltest twice and are missing CLI tests.

matheusaaguiar replied:
Fixed.

cameel commented:
I thought you went to sleep already so I actually also started a quick PR for this :) #16321

I got through yaml syntax errors, but still not sure if the tests will actually pass. If you're working on it now, please take a look and feel free to pull the missing parts into yours if you need them. I'll just focus on other things I wanted to do in the meantime.

matheusaaguiar commented:
> I thought you went to sleep already so I actually also started a quick PR for this :) #16321

I also thought you went to sleep and didn't check to see if there was already work in that direction :D

> I got through yaml syntax errors, but still not sure if the tests will actually pass. If you're working on it now, please take a look and feel free to pull the missing parts into yours if you need them. I'll just focus on other things I wanted to do in the meantime.

Ok, will do.

cameel commented on Changelog.md:
```diff
@@ -18,6 +18,7 @@ Bugfixes:
 
 Build System:
 * Enable Linux arm64 binaries for testing and releases.
+* Run soltest on arm64 binary.
```
W generally don't include CI changes in the changelog. I mean, adding a new architecture is notable and relevant to users, but no point mentioning new test runs.

matheusaaguiar commented:
> I'd check if the binary the build job saves to the workspace is under the same path as in the `b_ubu` job. I suspect it's not.
> 
> The soltest job is probably failing for a similar reason.

Yes, this is the reason. The build job saves the binaries with prefix `linux-arm`.
I will try to see if I can find a good way of adjusting for that.

cameel commented on .circleci/config.yml:
```diff
@@ -1293,6 +1298,21 @@ jobs:
+      - run:
+          name: Create symbolic link with expected soltest binary name
+          command: |
+            ln -s build/test/soltest-linux-arm build/test/soltest
```
I'd just change `persist_executables_to_workspace_arm` to keep it named `soltest`. Same for `solfuzzer`. it's only the `solc` binary that creates conflicts and needs a different name.

matheusaaguiar replied:
Changed.

cameel commented on .circleci/config.yml:
```diff
@@ -1355,6 +1375,16 @@ jobs:
+          name: Create symbolic link with expected solc binary name
+          command: |
+            ln -s build/solc/solc-static-linux-arm build/solc/solc
+      - cmdline_tests
```
Nice solution. We probably should do it with macOS as well (not necessarily in this PR). There we instead use `run_soltest` (which is an internal detail of the `soltest` command) because the executable name is different.

cameel commented on .circleci/config.yml:
```diff
@@ -1355,6 +1375,16 @@ jobs:
+      - run:
+          name: Create symbolic link with expected solc binary name
+          command: |
+            ln -s build/solc/solc-static-linux-arm build/solc/solc
```
```suggestion
            ln --symbolic build/solc/solc-static-linux-arm build/solc/solc
```

cameel reviewed (dismissed):
If the CI passes this is good enough, though it could also use a few small tweaks (see my comments). If you want to merge this already, they could be done in a separate PR as well.

r0qs commented on .circleci/config.yml:
```diff
@@ -1293,6 +1296,21 @@ jobs:
+    steps:
+      - run:
+          name: Make sure libevmone is found
+          comand: |
```
```suggestion
          command: |
```

cameel commented on .circleci/config.yml:
```diff
@@ -1293,6 +1296,21 @@ jobs:
+          name: Make sure libevmone is found
+          command: |
+            export PATH=$PATH:/usr/local/lib/
+      - soltest
```
Does the `export` actually do anything? I think it will be executed in a separate shell instance so `soltest` will not see it.

cameel replied:
The proper way to do this is to put the variable under `environment`. The problem is that you won't be able to access the existing value of `PATH` there.

What you could do instead might be to define an `evmone-path` parameter in the `soltest` command. Then have that command add it to `PATH` (if not empty).

nikola-matic replied:
There's no libevmone on the machine at all. See my comments above.

matheusaaguiar commented on .circleci/config.yml:
```diff
@@ -1355,6 +1373,16 @@ jobs:
+      - run:
+          name: Create symbolic link with expected solc binary name
+          command: |
+            ln --symbolic --relative build/solc/solc-static-linux-arm build/solc/solc
```
I think the problem here is that `checkout` and `attach_workspace` has not yet run and then there is no `build/solc` dir.

matheusaaguiar commented:
There are 3 smt tests which are failing:
`smtCheckerTests/external_calls/external_call_indirect_2.sol`
`smtCheckerTests/external_calls/external_call_indirect_5.sol`
`smtCheckerTests/functions/functions_external_2.sol`
 
The message changes from "violations happens" to "violation might happen", e.g. (smtCheckerTests/external_calls/external_call_indirect_2.sol):

```
Expected result:
  Warning 6328: (434-455): CHC: Assertion violation happens here.
  Warning 6328: (1270-1291): CHC: Assertion violation happens here.
  Info 1391: CHC: 2 verification condition(s) proved safe! Enable the model checker option "show proved safe" to see all of them.
Obtained result:
  Warning 6328: (434-455): CHC: Assertion violation happens here.
  Warning 6328: (1270-1291): CHC: Assertion violation might happen here.
  Info 1391: CHC: 2 verification condition(s) proved safe! Enable the model checker option "show proved safe" to see all of them.
```

Not sure if this happens because of ARM.
@blishko, can you take a look, please? What do you think?

blishko commented:
What I see now is that `soltest` is failing because it could not load evmone: `libevmone.so: cannot open shared object file: No such file or directory`.

I would need to see the failures.
It's possible that z3 behaves differently on arm.
Let me know when you fix the missing evmone problem.

matheusaaguiar committed:
Use new Linux ARM image

nikola-matic commented:
> @matheusaaguiar : #16330 is my proposed solution

Rebasing now, let's hope it helps.

cameel commented on .circleci/config.yml:
```diff
@@ -1355,6 +1382,24 @@ jobs:
+          name: Create symbolic link with expected solc and solfuzzer binaries name
+          command: |
+            ln --symbolic --relative build/solc/solc-static-linux-arm build/solc/solc
+            ln --symbolic --relative build/test/tools/solfuzzer-linux-arm build/test/tools/solfuzzer
```
Any reason https://github.com/argotorg/solidity/pull/16322#discussion_r2583703876 was reverted?

matheusaaguiar commented:
@cameel, We were getting this error (which @nikola-matic mentioned) when running bytecode compare:
```
Applying workspace layers
  0fff5a7b-a1ab-40c2-bf18-2531188c1769
  e6fd50ab-8077-45f8-b97b-3715cef3635b
  4cc2f52f-5512-4ef9-908b-81e6028a1956
  588dd45a-fe9d-4c8f-83d7-3e28f21a8657
Concurrent upstream jobs persisted the same file(s) into the workspace:
  - test/tools/solfuzzer
  - test/soltest

Error applying workspace layer for job 588dd45a-fe9d-4c8f-83d7-3e28f21a8657: Concurrent upstream jobs persisted the same file(s)
```

So I reverted https://github.com/argotorg/solidity/pull/16322#discussion_r2583703876 and updated the test jobs to use more fine-grained steps (instead of calling the templates `run_soltest` and `run_cmdline_tests`), so we can make symbolic links with the expected binaries names before running `soltest` and `cli` tests.
